### PR TITLE
Allow plugin create own deduplicate logic 

### PR DESCRIPTION
**Pull request by kuzmany**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" for all features, enhancements and bug fixes (until 3.3.0 is released) 
| Bug fix?                               | yes/no
| New feature?                           | yes/no
| Deprecations?                          | yes/no
| BC breaks?                             | yes/no
| Automated tests included?              | yes/no
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Follow up to https://github.com/mautic/mautic/pull/7502

Common `mautic:contacts:deduplicate` already deduplicate contacts by unique fields from configurations. But sometimes we need custom check for duplications (for example based on firstname/lastname/phone). This PR added dispatcher to checkForDuplicateContacts method allow load duplicate contacts by plugin. 
This allow also create own command (for example `mautic:contacts:deduplicate:custom`) along with core one. 

#### Steps to test this PR:
1. Install and setup MauticCustomDeduplicateBundle https://github.com/mtcextendee/mautic-custom-deduplicate-bundle
2. Choose in plugins settings:
firstname, lastname, phone
and not empty email
3. Create two contacts without email but with same firstname, lastname, and phone.
4. Run `php app/console mautic:contacts:deduplicate:custom`
5. See If contacts were merged
6. Then create two contacts again
7. Go to Mautic > Contacts
8. Run Custom deduplicate from menu
image
9. Wait and then see If you have at least two notification in notification area (started, finished)
image
10. Check If contacts were merged

### CODE REVIEW

File: app/bundles/LeadBundle/Event/DuplicateContactsEvent.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
+<?php
+
```
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DuplicateContactsEvent.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/LeadBundle/Event/DuplicateContactsEvent.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    public function __construct(array $fields)
+    {
+        $this->fields = $fields;
+    }
+
+    /**
+     * @return array
+     */
+    public function getFields()
```
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DuplicateContactsEvent.php:39]:
```suggestion
    public function getFields(): array
```
Please add param and return type for all new methods.

File: app/bundles/LeadBundle/Event/DuplicateContactsEvent.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * @var array
+     */
+    private $duplicates = [];
+
+    private $handledByPlugin = false;
+
+    /**
+     * CheckForDuplicateContactsEvent constructor.
+     */
```
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DuplicateContactsEvent.php:30]: There is no value in this comment.
```suggestion
```

File: app/bundles/LeadBundle/Exception/NotHandledDuplicationByPlugin.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ * @author      Mautic, Inc.
+ *
+ * @link        https://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Mautic\LeadBundle\Exception;
+
+class NotHandledDuplicationByPlugin extends \Exception
```
Comment (escopecz) [inline app/bundles/LeadBundle/Exception/NotHandledDuplicationByPlugin.php:14]:
```suggestion
class DuplicationNotHandledByPlugin extends \Exception
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Found some small improvements that can be made. Also please cover the rows that CodeCov is reporting with tests.

**Comment by donbowman** [comment]
seems this is still relevan to 4.0.0-rc since https://mtcextendee.com/our-works/mautic-plugins/custom-deduplicate-for-mautic/ would like to use it.

**Comment by RCheesley** [comment]
This one has a conflict to be resolved.

**Comment by kuzmany** [comment]
Let's close it. Do not support it anyomore

**Pull request closed without merging**